### PR TITLE
[BEAM-9424] Allow grouping by LogicalType

### PR DESCRIPTION
**Pull request by kanterov**
Add LogicalTypeCoder.

Support LogicalTypes in Group transform and SchemaCoder generation.

Related to: https://github.com/apache/beam/pull/11015

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by kanterov** [comment]
@reuvenlax it looks like there is a couple of places where LogicalType don't work since we introduced to/fromBaseType functions. Looking into the code we would need to support it in RowCoderGenerator that doesn't look straightforward because LogicalType doesn't have a static factory we can call.

Do you have thoughts on how we can fix that?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/RowCoderGenerator.java [hunk 1]
```diff
@@ -401,6 +404,20 @@ private static StackManipulation getCoder(Schema.FieldType fieldType) {
     }
   }
 
+  private static StackManipulation logicalTypeCoder(
+      Schema.LogicalType logicalType, StackManipulation baseCoder) {
+    throw new UnsupportedOperationException("not implemented");
```
Comment (kanterov) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/RowCoderGenerator.java]: This section is left unimplemented, and that's why the test is failing

**Comment by reuvenlax** [comment]
I'm not sure I understand the problem or the solution. Why don't logical types work in Group? Why does creating a new LogicalTypeCoder solve this?

**Comment by kanterov** [comment]
@reuvenlax because `RowCoderGenerator` uses `Coder<BaseT>`, not `Coder<InputT>`. We know how to produce `Coder<BaseT>`, and we can make `Coder<InputT>` out of it using `LogicalType`. So the test fails because Row builder expected `EnumerationType.Value` but got `String`. What is special about Group is that it constructs

**Comment by reuvenlax** [comment]
The whole point of Logical types is that the representation should be that of the base type (the schema returned by getBaseType). This means that the logical type can be processed by code that doesn't know about InputT, and also makes it possible to use it cross language. 

Where in Group do we need Coder<InputT>? I feel like I'm still missing something.

**Comment by kanterov** [comment]
@reuvenlax I cleaned the code and left only test that reproduces this issue.

The reason for that is because it uses `RowCoderGenerator` that would use a `Coder<BaseT>` instead of `Coder<InputT>` (see `RowCoderGenerator.java:374`). While it doesn't make any difference for binary representation, SDK code fails because of that. In the stack trace we can see how Combine should get `InputT=EnumerationType$Value`, but it gets `BaseT=Integer`.

```
Caused by: java.lang.ClassCastException: java.lang.Integer cannot be cast to org.apache.beam.sdk.schemas.logicaltypes.EnumerationType$Value
	at org.apache.beam.sdk.schemas.logicaltypes.EnumerationType.toBaseType(EnumerationType.java:37)
	at org.apache.beam.sdk.values.Row$Builder.verifyLogicalType(Row.java:659)
	at org.apache.beam.sdk.values.Row$Builder.verify(Row.java:652)
	at org.apache.beam.sdk.values.Row$Builder.verify(Row.java:635)
	at org.apache.beam.sdk.values.Row$Builder.build(Row.java:840)
	at org.apache.beam.sdk.schemas.utils.SelectHelpers.selectRow(SelectHelpers.java:194)
	at org.apache.beam.sdk.schemas.transforms.SchemaAggregateFn$Inner$ExtractSingleFieldFunction.apply(SchemaAggregateFn.java:241)
	at org.apache.beam.sdk.schemas.transforms.SchemaAggregateFn$Inner$ExtractSingleFieldFunction.apply(SchemaAggregateFn.java:231)
	at org.apache.beam.sdk.transforms.CombineFns$ComposedCombineFn.addInput(CombineFns.java:377)
	at org.apache.beam.sdk.schemas.transforms.SchemaAggregateFn$Inner.addInput(SchemaAggregateFn.java:283)
	at org.apache.beam.sdk.schemas.transforms.SchemaAggregateFn$Inner.addInput(SchemaAggregateFn.java:55)
	at org.apache.beam.sdk.transforms.Combine$CombineFn.apply(Combine.java:446)
	at org.apache.beam.sdk.transforms.Combine$GroupedValues$1.processElement(Combine.java:2171)
```

**Comment by reuvenlax** [comment]
@kanterov ok, I see the bug now. the problem is because we have an unfortunate asymmetry between Row.Builder.addValues and Row.getValues. getValues returns the base type for logical types (because that's how it's stored, but addValues expects to get the input type.

I'm still thinking of the best way to resolve this inconsistency without breaking other use cases. However a simple fix for now is to fix SelectHelpers to use attachValues instead of addValues. Coincidentally that change is already made in https://github.com/apache/beam/pull/11025. Can you check to see whether that PR fixes the problem you're seeing?

**Comment by reuvenlax** [comment]
BTW I think I need to prioritize fixing the inconsistency between getValues and addValues, as this is not the first time it's caused a bug.

**Comment by kanterov** [comment]
@reuvenlax I've rebased on the latest master that includes https://github.com/apache/beam/pull/11025. It didn't seem to help, because the test is failing with the same exception:

https://builds.apache.org/job/beam_PreCommit_Java_Commit/10259/testReport/junit/org.apache.beam.sdk.schemas.transforms/GroupTest/testAggregateLogicalType/

**Comment by reuvenlax** [comment]
@kanterov it's failing differently now - no longer in. SelectHelpers.

**Comment by reuvenlax** [comment]
@kanterov this is because (as I mentioned before) Row.getValue today always returns the base type (in this case int), because that's how we store it. This has the side effect of passing the base value into your combiner, which is currently expecting the value type.

This inconsistency keeps causing strange effects like this. I think we need getValue to return the same types passed into addValues (the value type). I'm trying to think of a good way to do this without breaking other semantics and impacting performance.

**Comment by kanterov** [comment]
@reuvenlax is there an option to have alternative to attachValues, but for getting values to be used in performance-critical code?

**Comment by reuvenlax** [comment]
IT's not just performance critical code. We wanted these transforms to be
usable by code that doesn't understand the logical type.

For example: you should be able to write a SQL expression over a type that
contains a OneOf. SQL doesn't understand the OneOf logical type of course,
however the base type is a Row with nullable fields, Since today getValue
returns the base type, you should be able to query it. Similar story with
Enumerations - you should be able to treat them as integers in a SQL
statement or in one of the schema transforms. (note: I"m not sure if this
has been tested with SQL, so it may not work today).

We need to figure out a good API that allows these transformation to use
the base type by default but _also_ allows you to write a combiner that
accepts the appropriate value type.

Reuven

**Pull request closed without merging**